### PR TITLE
Fix typeof Model errors throughout by using typeof Model generics

### PR DESCRIPTION
**Pull request by KapitanOczywisty**
Update of #868 with fixed lint errors. https://github.com/RobinBuschmann/sequelize-typescript/pull/868#issuecomment-774444691

cc: @Divlo

### CODE REVIEW

File: src/associations/belongs-to-many/belongs-to-many-association.ts [hunk 1]
```diff
@@ -8,28 +8,29 @@ import {ModelClassGetter} from "../../model/shared/model-class-getter";
...
     const options: OriginBelongsToManyOptions = {...this.options as any};
     const associatedClass = this.getAssociatedClass();
     const throughOptions = this.getThroughOptions(sequelize);
 
     const throughModel = typeof throughOptions === 'object' && typeof throughOptions.model !== "string" ? throughOptions.model : undefined;
     options.through = throughOptions;
-    options.foreignKey = getForeignKeyOptions(model, throughModel, this.options.foreignKey);
-    options.otherKey = getForeignKeyOptions(associatedClass, throughModel, this.options.otherKey);
+    options.foreignKey = getForeignKeyOptions(model, throughModel as any, this.options.foreignKey);
+    options.otherKey = getForeignKeyOptions(associatedClass, throughModel as any, this.options.otherKey);
```
Comment (Divlo) [inline src/associations/belongs-to-many/belongs-to-many-association.ts:33]: Is there is no way to avoid this `as any` type ?

**Review by Divlo — CHANGES_REQUESTED** [review]
Thanks for your contribution! @KapitanOczywisty 
That's great to have a stricter model defintion!

What if I don't want to provide `<TCreationAttributes, TModelAttributes>` is it still working ?
It should still works!

Also see my comment. :smile: 
We should avoid as much as possible `any` type, sometime we can't, we need to use `any`, we can't do something else.

File: src/associations/belongs-to-many/belongs-to-many-association.ts (hunk 1 above)
Comment (KapitanOczywisty) [reply src/associations/belongs-to-many/belongs-to-many-association.ts:33]: @Divlo You can see `any` in few other places in this file, seems like this was badly designed to begin with and it's beyond scope of this PR to fix this. I'm not seeing better fix without major redesign and getting modal attributes to work is way more important. Argument is not affecting return type anyway.

**Comment by KapitanOczywisty** [comment]
> What if I don't want to provide `<TCreationAttributes, TModelAttributes>` is it still working ?

In my testing both ways work now, seems like @intellix did a great job. When you don't provide attributes you have effectively `<{}, {}>`.

File: src/associations/belongs-to-many/belongs-to-many-association.ts (hunk 1 above)
Comment (Divlo) [reply src/associations/belongs-to-many/belongs-to-many-association.ts:33]: Right, maybe we'll need to refactor this code, someday.
I'll approve for the moment.

**Review by Divlo — APPROVED** [review]

**Review by RobinBuschmann — CHANGES_REQUESTED** [review]
Hey @KapitanOczywisty thanks for fixing this. Can you add some tests please?

**Comment by KapitanOczywisty** [comment]
@RobinBuschmann I probably could add a couple of types-only tests. What do you want to see in these tests?

**Comment by RobinBuschmann** [comment]
> @RobinBuschmann I probably could add a couple of types-only tests. What do you want to see in these tests?

The tests should validate that it's now possible to use `<TCreationAttributes, TModelAttributes>`. Yeah, checking the types should be enough 👍

**Comment by KapitanOczywisty** [comment]
Sooo, I see problems near everything what is touching `through` in `BelongsToMany` I have to rewrite this a bit.

And there are also problems with `FooModel.findAll({include: BarModel})`, but this is due to sequelize itself, so leave that for them to deal with?

**Comment by KapitanOczywisty** [comment]
@RobinBuschmann Fixed BelongsToMany and some testing added.

**Comment by KapitanOczywisty** [comment]
For userland type-guard have to be a bit more finicky and I'm thinking about exporting this one to userland, where you need to access some static properties from `Model`. I'd like to hear some feedback on that.

Generally, `type ModelType = Properties<typeof Model>` also works fine as type-guard and is fine when you don't care about attributes (which is the case in example bellow).

```ts
type Properties<T> = {
  [P in keyof T]: T[P];
}
type ModelType<TModelAttributes = any, TCreationAttributes = TModelAttributes> = Properties<typeof Model> & (new () => Model<TModelAttributes, TCreationAttributes>);

// ---

export function SequelizeFilter(model: ModelType, filterable: string[], filter?: string): WhereOptions | undefined {
  if (!filter || filter === "") return;

  const fields = [];
  for (let prop of filterable) {
    // rawAttributes works here because of Properties<typeof Model>
    // otherwise there would be: Property 'rawAttributes' does not exist on type 'ModelType<any, any>'.
    if (model.rawAttributes?.[prop]) {
      fields.push({ [prop]: { [Op.substring]: filter } });
    }
  }

  return fields.length ? { [Op.or]: fields } : undefined;
}
```

**Review by RobinBuschmann — APPROVED** [review]
Sorry for the late reply: Good work 👍

**Review by Divlo — APPROVED** [review]

**Comment by intellix** [comment]
Awesome, thanks for moving this forward :)

It seems my initial issues weren't addressed though in regards to Includeables: https://github.com/RobinBuschmann/sequelize-typescript/pull/868#issue-539045477

**Pull request merged**